### PR TITLE
(refactor, bugsbash)fix defer close statements

### PR DESCRIPTION
**Pull request by asquare14**
Signed-off-by: aSquare14 <atibhi.a@gmail.com>

Related to bugsbash

### CODE REVIEW

**Comment by akhilerm** [comment]
@asquare14 Can you check on the comments by the muse-bot

**Comment by asquare14** [comment]
@akhilerm Done ! Thanks for the review

**Comment by asquare14** [comment]
Can you please show me a screenshot of Better Code Hub, I am unable to view it.

File: pkg/mount/mountutil.go [hunk 1]
```diff
@@ -79,7 +80,17 @@ func (m DiskMountUtil) getDeviceMountAttr(fn getMountData) (DeviceMountAttr, err
 	if err != nil {
 		return mountAttr, err
 	}
-	defer file.Close()
+
+	defer func() {
```
Comment (akhilerm) [inline pkg/mount/mountutil.go:84]: The defer block will be executed only when the function is returned. So why is the `err` assignment done?

File: pkg/mount/mountutil.go [hunk 2]
```diff
@@ -168,7 +179,13 @@ func getRootPartition() (string, error) {
 	if err != nil {
 		return "", err
 	}
-	defer file.Close()
+
+	defer func() {
+		cerr := file.Close()
+		if cerr != nil {
+			err = cerr
```
Comment (akhilerm) [inline pkg/mount/mountutil.go:186]: Same comment as above

File: pkg/mount/mountutil.go [hunk 3]
```diff
@@ -180,7 +197,7 @@ func getRootPartition() (string, error) {
 		return "", err
 	}
 
-	return getDeviceName(link), nil
+	return getDeviceName(link), err
```
Comment (akhilerm) [inline pkg/mount/mountutil.go:200]: :+1:

File: pkg/udev/mockdata.go [hunk 4]
```diff
@@ -25,12 +25,13 @@ package udev
...
-	bd "github.com/openebs/node-disk-manager/blockdevice"
-	"github.com/openebs/node-disk-manager/pkg/mount"
-	"github.com/openebs/node-disk-manager/pkg/sysfs"
 	"io/ioutil"
 	"os"
 	"strings"
+
+	bd "github.com/openebs/node-disk-manager/blockdevice"
+	"github.com/openebs/node-disk-manager/pkg/mount"
+	"github.com/openebs/node-disk-manager/pkg/sysfs"
```
Comment (akhilerm) [inline pkg/udev/mockdata.go:34]: :+1:

**Review by akhilerm — COMMENTED** [review]
Given some comments.

**Comment by akhilerm** [comment]
> Can you please show me a screenshot of Better Code Hub, I am unable to view it.

@asquare14 The Better code hub failure is expected. Its not because of changes in this PR. Only because the file was evaluated now, the issues from previous changes came in.

File: pkg/mount/mountutil.go (hunk 1 above)
Comment (asquare14) [reply pkg/mount/mountutil.go:84]: Fixed

File: pkg/mount/mountutil.go [hunk 5]
```diff
@@ -79,7 +80,15 @@ func (m DiskMountUtil) getDeviceMountAttr(fn getMountData) (DeviceMountAttr, err
...
 	}
-	defer file.Close()
+
+	defer func() {
+		file.Close()
+	}()
+
+	if err != nil {
+		return mountAttr, err
+	}
```
Comment (akhilerm) [inline pkg/mount/mountutil.go:90]: This error check is not needed here.

File: pkg/mount/mountutil.go [hunk 6]
```diff
@@ -168,7 +177,10 @@ func getRootPartition() (string, error) {
 	if err != nil {
 		return "", err
 	}
-	defer file.Close()
+
+	defer func() {
+		file.Close()
+	}()
```
Comment (akhilerm) [inline pkg/mount/mountutil.go:183]: @asquare14 What is the advantage of using the above format for `defer` compared to
```golang
defer file.Close()
```

Comment (richardelling) [reply pkg/mount/mountutil.go:183]: file.Close() can fail and you need to check the return code. This is not a problem for files opened for reading, but can be  a big problem for files opened for writing. The automated code checkers aren't very smart about this, so they assume the worst: the file is open for writing and you need to deal with failed Close(). 

NB, the above code still ignores the error condition, but it fakes out the automated code checker. Perhaps a better solution is to either explicitly ignore the error condition in the defer func for files opened for read or add a comment so the code checker ignores that check for this instance.

Comment (akhilerm) [reply pkg/mount/mountutil.go:183]: @asquare14 Can you make the necessary changes as per the above comment. Here, the file is opened for reading. So,  we can explicitly ignore the error.

Comment (richardelling) [reply pkg/mount/mountutil.go:183]: what I mean by explicitly ignoring for files that are opened for reading is something like:
```
defer func() {
    _ := file.Close()
}
```

**Comment by akhilerm** [comment]
@asquare14 Can you address the review comments on the PR.

**Pull request closed without merging**